src/regex/pcre2.cr

```diff
@@ -137,27 +142,45 @@ module Regex::PCRE2
     end
   end
 
-  class_getter general_context do
-    LibPCRE2.general_context_create(->(size, _data) { GC.malloc(size) }, ->(pointer, _data) { GC.free(pointer) }, nil)
+  class_getter match_context : LibPCRE2::MatchContext* do
+    match_context = LibPCRE2.match_context_create(nil)
+    LibPCRE2.jit_stack_assign(match_context, ->(_data) { Regex::PCRE2.jit_stack }, nil)
+    match_context
   end
 
   # Returns a JIT stack that's shared in the current thread.
   #
   # Only a single `match` function can run per thread at any given time, so there
   # can't be any concurrent access to the JIT stack.
-  @@jit_stack = Crystal::ThreadLocalValue(LibPCRE2::JITStack).new
+  @@jit_stack = Crystal::ThreadLocalValue(LibPCRE2::JITStack*).new
 
   def self.jit_stack
     @@jit_stack.get do
-      LibPCRE2.jit_stack_create(32_768, 1_048_576, general_context) || raise "Error allocating JIT stack"
+      LibPCRE2.jit_stack_create(32_768, 1_048_576, nil) || raise "Error allocating JIT stack"
+    end
+  end
+
+  # Match data is shared per instance and thread.
+  #
+  # Match data contains a buffer for backtracking when matching in interpreted mode (non-JIT).
+  # This buffer is heap-allocated and should be re-used for subsequent matches.
+  @match_data = Crystal::ThreadLocalValue(LibPCRE2::MatchData*).new
+
+  private def match_data
+    @match_data.get do
+      LibPCRE2.match_data_create_from_pattern(@re, nil)
+    end
+  end
+
+  def finalize
+    @match_data.get?.try do |match_data|
```

Comment:
IIUC this is called exactly once when the object is unreachable from all threads, so freeing just the match data for the current thread isn't sufficient. This should probably use `@match_data.consume_each` or even `@match_data.@values.each_value`

Reply:
Good catch 👍